File: parth/Week2/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::vector<std::vector<int>> sol;
        std::sort(nums.begin(), nums.end());
        for(int i=0; i<nums.size(); i++)
        {
            for(int j=i+1,k=nums.size()-1;j<k;)
            {
                long sum = nums[j] + nums[k] + nums[i];
                if(sum == 0)
                {
                    std::vector<int> triplet;
                    triplet.push_back(nums[i]);
                    triplet.push_back(nums[j]);
                    triplet.push_back(nums[k]);
                    sol.push_back(triplet);
                    j++;
                    k--;
                    while(j<nums.size() && nums[j] == nums[j-1])
                    {
                        j++;
                    }
                    while(k>=0 && k+1<nums.size() && nums[k] == nums[k+1])
                    {
                        k--;
                    }
                } 
                else if(sum > 0)
                {
                    k--;
                } 
                else if (sum < 0)
                {
                    j++;
                }
            }
            while(i+1 < nums.size() && nums[i] == nums[i+1])
            {
                i++;
            }
        }
        return sol;
    }
};

File: Manali/Week3/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        found = set()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            target = -nums[i]
            d = set()
            for j in range(i+1,len(nums)):
                if target-nums[j] in d:
                    l = [nums[i], nums[j], target-nums[j]]
                    if tuple(l) not in found:
                        ans.append(l)
                        found.add(tuple(l))
                else:
                    d.add(nums[j])
        return ans